streamtools/consumers.py

```diff
@@ -74,6 +78,48 @@ def __call__(self, queue_arg="", override=False, *args, **kwargs):
 
         return self._decorator(set_queue_label, *args, **kwargs)
 
+    def parse_decorated_args(self, func, args_list, kwargs_list):
+        args_error_msg = "Please use only one argument (and optional 'self' arg in classes)" + \
+                         " for the message to be passed."
+
+        if kwargs_list:
+            kw_error_msg = "No keyword arguments allowed. "
+            raise ValueError(args_error_msg + kw_error_msg)
+        if len(args_list) > 1:
+            raise ValueError(args_error_msg)
+
+        self_obj = None
+        if args_list:
+            self_obj, *_ = args_list
+
+            self_members = dict(inspect.getmembers(self_obj.__class__))
+            func_in_class = self_members.get(func.__name__)
+            assert func_in_class and \
+                   inspect.isroutine(func) and \
+                   (func.__qualname__ == func_in_class.__qualname__), \
+                "First argument must be either 'self/cls' inside a class object. " + \
+                "Only one argument allowed otherwise for unbound functions."
+
+        # Save 'self_obj' to consumer class instance on first run
+        if self.in_class is None:
+            self.self_obj = self_obj
+            self.in_class = True if args_list else False
+
+        return self_obj
+
+    async def call_decorated(self, func, msg, w_args, w_kwargs):
+        self_obj = self.parse_decorated_args(func, w_args, w_kwargs)
+        if self.in_class:
+            try:
+                assert self_obj == self.self_obj, \
```

Comment:
Really cool way to if/else (reject), still catches me to see an `assert` in functional code instead of just tests 👆

Reply:
Oh ya, hadn't even realised that's what I did here 🙈  ... this could've also easily been
```
if not self_obj == self.self_obj:
    log.error("'self/cls' argument not being consistently passed into class method")
    await func(self.self_obj, msg)
else:
    await func(self_obj, msg)
```

This reminds me of [this Chris Neugebauer talk](https://youtu.be/mBavPw0Kc0w?t=813) where he did some wild things with control flow using the `raise` keyword and `BaseException` class in Python